mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/MlflowAutologEventPublisher.scala

```diff
@@ -0,0 +1,130 @@
+package org.mlflow.spark.autologging
+
+import java.util.concurrent._
+
+import org.apache.spark.sql.SparkSession
+import org.slf4j.LoggerFactory
+import py4j.Py4JException
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+/**
+  * Object exposing the actual implementation of MlflowAutologEventPublisher.
+  * We opt for this pattern (an object extending a trait) so that we can mock methods of the
+  * trait in testing
+  */
+object MlflowAutologEventPublisher extends MlflowAutologEventPublisherImpl {
+
+}
+
+/**
+ * Trait implementing a publisher interface for publishing events on Spark datasource reads to
+ * a set of listeners. See the design doc:
+ * https://docs.google.com/document/d/11nhwZtj-rps0stxuIioFBM9lkvIh_ua45cAFy_PqdHU/edit for more
+ * details.
+ */
+private[autologging] trait MlflowAutologEventPublisherImpl {
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  private[autologging] var sparkQueryListener: SparkDataSourceListener = _
+  private val ex = new ScheduledThreadPoolExecutor(1)
+  private[autologging] var subscribers: mutable.LinkedHashMap[String, MlflowAutologEventSubscriber] =
+    mutable.LinkedHashMap[String, MlflowAutologEventSubscriber]()
+  private var scheduledTask: ScheduledFuture[_] = _
+
+  def spark: SparkSession = {
+    SparkSession.builder.getOrCreate()
```

Comment:
Should we call get or create here or just try to access the existing session and fail if it does not exist? In other places I have opted for the latter so that we do not change the state in an unexpected way.